File: src/main/java/ca/rafemurray/CourseMarkerItem.java
package ca.rafemurray;

import net.minecraft.item.Item;
import net.minecraft.item.ItemUsageContext;
import net.minecraft.util.ActionResult;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;

public class CourseMarkerItem extends Item {
    private final Course course;
    public CourseMarkerItem(Settings settings) {
        super(settings);
        this.course = new Course();
    }
    //TODO: update the course fields for the start and finish blocks

    @Override
    public ActionResult useOnBlock(ItemUsageContext context) {
        World world = context.getWorld();
        BlockPos pos = context.getBlockPos();
        if (world.getBlockEntity(pos) instanceof StartBlockEntity startBlockEntity){
            course.setStart(startBlockEntity);
            startBlockEntity.setCourse(course);
            return ActionResult.SUCCESS;
        }
        if (world.getBlockEntity(pos) instanceof FinishBlockEntity finishBlockEntity) {
            course.setFinish(finishBlockEntity);
            finishBlockEntity.setCourse(course);
            return ActionResult.SUCCESS;
        }
        return ActionResult.PASS;
    }
}

File: src/main/java/ca/rafemurray/item/CourseMarkerItem.kt
package ca.rafemurray.item

import ca.rafemurray.block.entity.FinishBlockEntity
import ca.rafemurray.block.entity.StartBlockEntity
import ca.rafemurray.course.Course
import net.minecraft.world.InteractionResult
import net.minecraft.world.item.Item
import net.minecraft.world.item.context.UseOnContext

class CourseMarkerItem(settings: Properties) : Item(settings) {

    private val course: Course = Course()

    override fun useOn(context: UseOnContext): InteractionResult {
        val level = context.level
        val pos = context.clickedPos

        val blockEntity = level.getBlockEntity(pos)
        if (blockEntity is StartBlockEntity) {
            course.setStart(pos)
            blockEntity.setCourse(course)
            return InteractionResult.SUCCESS
        }
        if (blockEntity is FinishBlockEntity) {
            course.setFinish(pos)
            blockEntity.setCourse(course)
            return InteractionResult.SUCCESS
        }
        return InteractionResult.PASS
    }
}
